packages/walletconnect/src/index.ts

```diff
@@ -75,47 +83,40 @@ export class WalletConnect extends Connector {
     if (error) this.onError?.(error)
   }
 
-  private chainChangedListener = (chainId: number | string): void => {
-    this.actions.update({ chainId: parseChainId(chainId) })
+  private chainChangedListener = (chainId: string | number): void => {
+    // @todo incorrect type in WC, this is always `number`
+    this.actions.update({ chainId: +chainId })
   }
 
   private accountsChangedListener = (accounts: string[]): void => {
     this.actions.update({ accounts })
   }
 
-  private URIListener = (_: Error | null, payload: { params: string[] }): void => {
-    this.events.emit(URI_AVAILABLE, payload.params[0])
+  private URIListener = (uri: string): void => {
+    this.events.emit(URI_AVAILABLE, uri)
   }
 
-  private async isomorphicInitialize(chainId = this.defaultChainId): Promise<void> {
-    if (this.eagerConnection) return
-
-    // because we can only use 1 url per chainId, we need to decide between multiple, where necessary
-    const rpc = Promise.all(
-      Object.keys(this.rpc).map(
-        async (chainId): Promise<[number, string]> => [
-          Number(chainId),
-          await getBestUrl(this.rpc[Number(chainId)], this.timeout),
-        ]
-      )
-    ).then((results) =>
-      results.reduce<{ [chainId: number]: string }>((accumulator, [chainId, url]) => {
-        accumulator[chainId] = url
-        return accumulator
-      }, {})
-    )
-
-    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (m) => {
-      this.provider = new m.default({
+  private isomorphicInitialize(
+    desiredChainId: number | undefined = this.defaultChainId
+  ): Promise<WalletConnectProvider> {
+    if (this.eagerConnection) return this.eagerConnection
+
+    const rpcMap = this.rpcMap ? getBestUrlMap(this.rpcMap, this.timeout) : undefined
+    const chains = desiredChainId ? getChainsWithDefault(this.chains, desiredChainId) : this.chains
+
+    return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (ethProviderModule) => {
+      const provider = (this.provider = await ethProviderModule.default.init({
         ...this.options,
-        chainId,
-        rpc: await rpc,
-      }) as unknown as MockWalletConnectProvider
-
-      this.provider.on('disconnect', this.disconnectListener)
-      this.provider.on('chainChanged', this.chainChangedListener)
-      this.provider.on('accountsChanged', this.accountsChangedListener)
-      this.provider.connector.on('display_uri', this.URIListener)
+        chains,
+        rpcMap: await rpcMap,
+      }))
+
+      provider.on('disconnect', this.disconnectListener)
+      provider.on('chainChanged', this.chainChangedListener)
+      provider.on('accountsChanged', this.accountsChangedListener)
+      provider.on('display_uri', this.URIListener)
+
+      return provider
```

Comment:
I think it's confusing that you `return provider`, instead of `void`, because part of `isomorphicInitialize` is setting `this.provider`, so you now have multiple ways of accessing the provider.

Would you mind changing the signature to return `Promise<void>`, and always awaiting and then accessing the provider through `this.provider`?

Reply:
Yup, I updated occurrences inside `activate` as per discussion on Slack. Changing signature back to `Promise<void>` is likely to introduce another equally confusing maintenance challenge that suggests `this.provider` could be undefined after initialising it. Explicit return does solve this problem.
<img width="424" alt="Screenshot 2023-02-18 at 01 22 30" src="https://user-images.githubusercontent.com/2464966/219798070-a72ec787-0af4-42cc-b7ff-4a962e7167c0.png">